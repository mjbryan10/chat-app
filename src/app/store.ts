import { configureStore, ThunkAction, Action } from '@reduxjs/toolkit';
import loginReducer from '../features/login/loginSlice';
import usersReducer from '../features/users/usersSlice';
import themeReducer from '../features/theme/themeSlice';
import conversationReducer from '../features/conversation/conversationSlice';
import messageReducer from '../features/message/messageSlice';
import navigationReducer from '../features/navigation/navigationSlice'

export const store = configureStore({
  reducer: {
    login: loginReducer,
    users: usersReducer,
    theme: themeReducer,
    conversation: conversationReducer,
    message: messageReducer,
    navigation: navigationReducer,
  },
});

export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;

/**
 * String values available to the loading status.
 */
export type LoadingStatus = 'rejected' | 'pending' | 'fulfilled' | 'idle';
