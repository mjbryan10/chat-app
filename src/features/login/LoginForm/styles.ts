import styled, { ThemeProps, Theme } from 'styled-components';

export const Container = styled.div`
   position: fixed;
   top: 0;
   right: 0;
   bottom: 0;
   left: 0;
   flex-flow: row nowrap;
   align-items: center;
   justify-content: center;
   display: flex;
   background: ${({ theme }: ThemeProps<Theme>) => theme.background.page ?? 'black'};
`;

export const Form = styled.form`
   display: flex;
   flex-flow: column nowrap;
   padding: 1em;
   position: relative;
   width: 100%;
   max-width: 400px;
   & input {
      margin-bottom: 3px;
   }
   & input[type='submit'] {
      margin-top: 5px;
   }
`;

export const ErorrText = styled.p<ThemeProps<Theme>>`
   color: ${({theme})=> theme.color.danger || 'red'};
`;
