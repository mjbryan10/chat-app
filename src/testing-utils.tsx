import React from 'react';
import { render } from '@testing-library/react'
import { ThemeProvider } from 'styled-components'
import { dark } from './shared/theme';

const AllTheProviders = ({ children }: any) => {
  return (
    <ThemeProvider theme={dark}>
        {children}
    </ThemeProvider>
  )
  
}

const customRender = (ui?: any, options?: any) =>
  render(ui, { wrapper: AllTheProviders, ...options })

// re-export everything
export * from '@testing-library/react'

// override render method
export { customRender as render }