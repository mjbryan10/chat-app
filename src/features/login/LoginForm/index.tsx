import React, { useState, FC } from 'react';
import InputTextFieldWithLabel from '../../../components/InputTextFieldWithLabel';
import InputSubmitButton from '../../../components/InputSubmitButton';
import * as S from './styles';
import { useSelector } from 'react-redux';
import { selectLogin } from '../loginSlice';

interface Props {
   id: number;
   name: string;
   handleSubmit?: () => void;
}

/**
 * A React functional component for rendering a login form.
 * @note As this production is for demo purposes, the login
 * credentials are being passed as props. In future they
 * should been handled within the component.
 * @param id The users ID
 * @param name The username
 * @param handleSubmit The submit handler that will be called on form submission
 */
const LoginForm: FC<Props> = ({ id, name, handleSubmit }) => {
   //REDUX:
   const { isAuthenticated } = useSelector(selectLogin);

   //STATE: @see note above.
   const [idState, setIdState] = useState(id);
   const [nameState, setNameState] = useState(name);

   //METHODS:
   const onSubmit = (event: React.FormEvent) => {
      event.preventDefault();
      if (handleSubmit) handleSubmit();
   };

   return (
      <S.Form action="submit" onSubmit={onSubmit} data-testid="login-form">
         <InputTextFieldWithLabel
            handleChange={(value: string) => setNameState(value)}
            labelValue="Name"
            initialValue={nameState}
            disabled
         />
         <InputTextFieldWithLabel
            handleChange={(value: string) => setIdState(parseInt(value))}
            labelValue="ID"
            initialValue={idState.toString()}
            disabled
         />
         {isAuthenticated === false ? (
            <S.ErorrText>
               It appears we were unable to log you in with those details, please try
               again later.
            </S.ErorrText>
         ) : null}
         <InputSubmitButton value="Log in" />
      </S.Form>
   );
};

export default LoginForm;
