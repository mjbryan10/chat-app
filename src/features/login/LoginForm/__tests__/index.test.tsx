import React from 'react';
import LoginForm from '..';
import { render, fireEvent, cleanup } from 'testing-utils';
import axios from 'axios';
import { Provider } from 'react-redux';
import { store } from '../../../../app/store';

jest.mock('axios');

describe('Login Component', () => {
   const __user = { id: 1, name: 'wessel' };
   const __handleSubmit = jest.fn();
   beforeEach(() => {
      (axios.get as jest.Mock).mockClear();
   });

   afterEach(() => {
      cleanup()
      __handleSubmit.mockClear();
   });

   const setup = () => {
      const utils = render(
         <Provider store={store}>
            <LoginForm id={__user.id} name={__user.name} handleSubmit={__handleSubmit} />
         </Provider>
      );
      const nameInput = utils.getAllByTestId('input-text-field')[0];
      const idInput = utils.getAllByTestId('input-text-field')[1];
      const submitButton = utils.getByTestId('input-submit-button');
      const form = utils.getByTestId('login-form');
      return { nameInput, idInput, submitButton, form, ...utils };
   };

   test('should call submit handler if form is submitted', () => {
      const { form } = setup();

      fireEvent.submit(form);

      expect(__handleSubmit).toBeCalledTimes(1);
   });

   test('should trigger submit when `Log In` button is clicked', () => {
      const { submitButton } = setup();

      fireEvent.click(submitButton);

      expect(__handleSubmit).toBeCalledTimes(1);
   });

   test('should render correctly', () => {
   const { asFragment } = setup();

   expect(asFragment()).toMatchSnapshot();
});
});
