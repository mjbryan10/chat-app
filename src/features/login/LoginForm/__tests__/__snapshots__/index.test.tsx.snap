// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Component should render correctly 1`] = `
<DocumentFragment>
  <form
    action="submit"
    class="sc-AxhUy hyKAMG"
    data-testid="login-form"
  >
    <label
      class="sc-AxirZ jSgOQo"
    >
      Name
      <input
        class="sc-AxjAm kuGIAZ"
        data-testid="input-text-field"
        placeholder=""
        value="wessel"
      />
    </label>
    <label
      class="sc-AxirZ jSgOQo"
    >
      ID
      <input
        class="sc-AxjAm kuGIAZ"
        data-testid="input-text-field"
        placeholder=""
        value="1"
      />
    </label>
    <input
      class="sc-AxiKw joUDrS"
      data-testid="input-submit-button"
      type="submit"
      value="Log in"
    />
  </form>
</DocumentFragment>
`;
